### PR TITLE
Fixing equilibration when having numerical aquifers.  

### PR DESCRIPTION
**Pull request by GitPaean**
With the targeting three phase black oil case, the reservoir ZCORN range is between 2000 and 1000, and numerical aquifer cell depth is 2400. 

For one of the equilibration regions, the reservoir `vspan` based on `cellZMinMax_`  is [2000, 2002],  water-oil contact depth is 2003, oil-gas contact depth is 1500.  At the end, the `vspan` used for `ptable` equilibration is [1500, 2003]. 

At the end, when `makeWatPressure`, ` this->value_[Direction::Down]` is a table between [2003, 2003]. As a result, for any depth bigger than 2003, the water pressure will be `nan` based on extrapolation. For example, the numerical aquifer cell has a depth 2040, it gets `nan` water pressure value. 

There are different ways to fix this. The equilibration code itself can be done differently when considering the depth range for the pressure table. 

The PR tries not to touch the general equilibration code and limit the impact within the numerical aquifer related cases. I think this can be the minimal change to tangle the problem while hopefully not introducing new problems.

### CODE REVIEW

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by GitPaean** [comment]
I did not reproduce the test failure from jenkins. I am using update_data as a tool to see what exactly happens.  anyway, the difference reported in jenkins is not worrying at all, just my local tests does not show any difference for those cases. 

`update_data` is not preparing for merging.

**Comment by GitPaean** [comment]
jenkins build this update_data please

**Comment by GitPaean** [comment]
I believe this PR can fix the running of the aquifer cases. 

But at the same time, the equilibration code itself might still be able to run into the same issue under some specific setup circumstances. Although not thinking very thoroughly, as long as the equilibration  oil-water contact depth is deeper than the deepest `cellZMinMax_` for the equlibration region, we will create a water pressure table that is not able to do extrapolation.  I think it only causes problem for the cases that modify depth of the cell manually. As long as we do not update the `cellZMinMax_` when we update the depth, we might run into the same problem. 

Anyway, after phrasing this message, I think it is a okay fix for the targeting numerical aquifer case. We just need to confirm the running results is okay, then this PR can go in.

**Comment by GitPaean** [comment]
Checked the update_data output, it is safe to update the reference results, waiting for the confirmation from the user side. 
```
summary.x 3D_2AQU_NUM. ANQR:2 ANQT:2
-- Days   dd/mm/yyyy             ANQR:2           ANQT:2 
---------------------------------------------------------
   1.00   02/11/2018             1.4608           1.4608 
   4.00   05/11/2018            44.6161          135.309 
  13.00   14/11/2018            307.997          2907.28 
  30.00   01/12/2018            774.689            16077 
  61.00   01/01/2019            1273.12          55543.8 
 120.00   01/03/2019            1685.53           154990 
 151.00   01/04/2019            1817.95           211346 
 242.00   01/07/2019            1947.43           388562 
 304.00   01/09/2019            1986.91           511751 
```
```
summary.x 3D_2AQU_NUM. ANQR:2 ANQT:2
-- Days   dd/mm/yyyy             ANQR:2           ANQT:2 
---------------------------------------------------------
   1.00   02/11/2018            1.64716          1.64716 
   4.00   05/11/2018            44.6034          135.457 
  13.00   14/11/2018            307.982           2907.3 
  30.00   01/12/2018            774.677          16076.8 
  61.00   01/01/2019            1273.11          55543.4 
 120.00   01/03/2019            1685.52           154989 
 151.00   01/04/2019            1817.95           211346 
 242.00   01/07/2019            1947.43           388562 
 304.00   01/09/2019            1986.91           511750 
```

**Comment by GitPaean** [comment]
The result for the targeting case revealed a small pressure difference with the patch against the reference result. about `0.006` bar.  As a result, the WBHP for the injector and producer also differ with similar amount. At the same time, the aquifer rate is very  small, this small pressure difference affects the aquifer rate in a significant way (relative). 

Will investigate and discuss whether this is a problem.

**Comment by GitPaean** [comment]
With some analysis, I think the pressure difference of `0.006` bar is at the numerical error range should not cause concerns. It needs to be agreed before we get this PR in.

**Comment by GitPaean** [comment]
The results are accepted. I am making the PR ready for review. With approval of the reviewers, we prepare to get this PR in.

**Comment by GitPaean** [comment]
jenkins build this opm-tests=757 please

**Comment by GitPaean** [comment]
Not sure about the testing failure for the parallel runnning, will check. 

```
201/205 Test #194: compareParallelSim_flow+3D_2AQU_NUM ...............................***Failed    1.72 sec
```

**Comment by GitPaean** [comment]
I do not reproduce the difference locally for the running with 1 process and 4 processes, 

```
opm-tests/aquifer-num/devel3$ summary.x 3D_2AQU_NUM. ANQP:1                   ANQP:2                   ANQR:1                   ANQR:2                   ANQT:1  ANQT:2
-- Days   dd/mm/yyyy             ANQP:1           ANQP:2           ANQR:1           ANQR:2           ANQT:1           ANQT:2 
-----------------------------------------------------------------------------------------------------------------------------
   1.00   02/11/2018            258.128          258.128         0.144861          1.64716         0.144861          1.64716 
   4.00   05/11/2018            258.128          258.128          3.92355          44.6034          11.9155          135.457 
  13.00   14/11/2018            258.127          258.117          27.1961          307.982           256.68           2907.3 
  30.00   01/12/2018            258.122          258.069          68.8935          774.677          1427.87          16076.8 
  61.00   01/01/2019            258.109          257.925          114.849          1273.11          4988.18          55543.4 
 120.00   01/03/2019            258.073          257.562          156.596          1685.52          14227.3           154989 
 151.00   01/04/2019            258.053          257.356           171.36          1817.95          19539.5           211346 
 242.00   01/07/2019            257.985          256.708           192.55          1947.43          37061.6           388562 
 304.00   01/09/2019            257.936          256.258           202.82          1986.91          49636.4           511750 

opm-tests/aquifer-num/devel3-4$ summary.x 3D_2AQU_NUM. ANQP:1                   ANQP:2                   ANQR:1                   ANQR:2                   ANQT:1  ANQT:2
-- Days   dd/mm/yyyy             ANQP:1           ANQP:2           ANQR:1           ANQR:2           ANQT:1           ANQT:2 
-----------------------------------------------------------------------------------------------------------------------------
   1.00   02/11/2018            258.128          258.128         0.143251          1.62885         0.143251          1.62885 
   4.00   05/11/2018            258.128          258.128          3.93046          44.6821          11.9346          135.675 
  13.00   14/11/2018            258.127          258.117          27.1835          307.839          256.586          2906.23 
  30.00   01/12/2018            258.122          258.069          68.8955          774.701          1427.81          16076.1 
  61.00   01/01/2019            258.109          257.925          114.849          1273.12          4988.13          55542.8 
 120.00   01/03/2019            258.073          257.562          156.596          1685.53          14227.3           154989 
 151.00   01/04/2019            258.053          257.356          171.365             1818          19539.6           211347 
 242.00   01/07/2019            257.985          256.708          192.551          1947.44          37061.7           388564 
 304.00   01/09/2019            257.936          256.258          202.819           1986.9          49636.5           511751 

```

**Comment by GitPaean** [comment]
jenkins build this update_data please

**Comment by bska** [comment]
> I do not reproduce the difference locally for the running with 1 process and 4 processes,

Did you try
```sh
compareECL -a -t SMRY devel3{,-4}/3D_2AQU_NUM 0.02 8.0e-5
```
which is the way that the Jenkins CI system runs the comparisons?

**Comment by GitPaean** [comment]
> > I do not reproduce the difference locally for the running with 1 process and 4 processes,
> 
> Did you try
> 
> ```shell
> compareECL -a -t SMRY devel3{,-4}/3D_2AQU_NUM 0.02 8.0e-5
> ```
> 
> which is the way that the Jenkins CI system runs the comparisons?

I think for this specific case, the Jenkins CI is using the following parameters, 

```
compareECL -a -t SMRY devel3{,-4}/3D_2AQU_NUM 0.12 0.01
```

With these relaxed parameters, local running passed the test, but it can be just unlucky from the Jenkins system.  Like if we changed the relative tolerance to be 0.0104, it will pass. 

Anyway, we should also have a understanding about why the running results for this specific case changes at all.  Will report back tomorrow and also discuss how to make Jenkins happy.

**Comment by GitPaean** [comment]
I hope I did not do anything wrong. 

For the serial running test against the ECL file. 
Locally, the master branch, the branch with this PR and the opm-simulation-reference/flow have the exactly the same output. 

Somehow, Jenkins thinks differently, and a update of the reference is required for the case ` 3D_2AQU_NUM`. Locally, it is not necessary and there should be no change of the results due to this PR for the serial running. 

For the parallel running with 4 processes, 
The results for the master and the results from this PR are different, and they are both different from the serial running results. 

But both can pass the test 
```
compareECL -a -t SMRY devel3{,-4}/3D_2AQU_NUM 0.12 0.01
```

Somehow, the jenkins' comparison is a little off due to the relative tolerance. 

To sum up, 
1. Jenkins used to produce the same results with my local master branch.  The serial running with this PR does not have any difference from the master branch (which should be), and Jenkins produces different serial running results with this PR and thinks `update_data` is needed. 
2. The parallel running (4 pcs) with master and this PR both produces different results from the serial running. And they are also different from each other.  Not sure `zSpan` and `cellZMinMax` are used in the parallel running.  The parallel running regression test failed due to exceed the relative tolerance with a small margin.

**Comment by GitPaean** [comment]
I will increase the relative tolerance for this specific case parallel regression test from 0.1 to 0.11 to make it pass and let reviewers to determine whether it is a appropriate.

**Comment by GitPaean** [comment]
I am increasing the absolute tolerance for the parallel regression test of this specific case from `0.12` to `0.17` to make the test pass and leave to the reviewers to judge.  I personally do not like it while I just do not have other ideas to go. 

Especially for this case, Jenkins produces same serial running result with my local machine, and produces different results with this PR. 
My locally running indicates that there is no change in the serial running with this PR, which is the expected behavoir.

**Comment by GitPaean** [comment]
jenkins build this opm-tests=759 please

**Comment by blattms** [comment]
>     2. The parallel running (4 pcs) with master and this PR both produces different results from the serial running. And they are also different from each other.  Not sure `zSpan` and `cellZMinMax` are used in the parallel running.  The parallel running regression test failed due to exceed the absolute tolerance for the first time step.

So this means that the results between serial and parallel were different before the change and not are different in another way. Might be a hint that there has been a parallel regression before that went unnoticed.

**Comment by GitPaean** [comment]
> So this means that the results between serial and parallel were different before the change and now are different in another way.

as far as I remember, it is correct.

**Comment by bska** [comment]
> > So this means that the results between serial and parallel were different before the change and now are different in another way.
> 
> as far as I remember, it is correct.

Hi @GitPaean: What does "it" refer to here?  @blattms's statement?  The new source code?  Something else?

**Comment by GitPaean** [comment]
> Hi @GitPaean: What does "it" refer to here? @blattms's statement? The new source code? Something else?

It refers to ` @blattms's statement?`

**Comment by totto82** [comment]
jenkins build this opm-tests=759 please

**Comment by totto82** [comment]
Any reasons for not merging this? The code look good and it fix a bug in the code. I restarted Jenkins to assure the reference case update has not be polluted.

**Comment by GitPaean** [comment]
I mean, it is good if we do not need to update the reference result when unnecessary.

**Comment by bska** [comment]
> @bska , do you have some opinions here

I do.  It's **probably** related to the fact that we currently don't integrate the same vertical span on all processes.  Adding more disparity on those ranks that happen to have an aquifer will only serve to amplify the problem.  That said, I don't have a better suggestion right now and if we still continue to produce reasonable numerical solutions then I think this is an acceptable stopgap measure.  We will have to revisit the problem before the next release though, because I'm not really comfortable having a system that produces different solutions when run in parallel versus when run in sequential mode.

**Comment by GitPaean** [comment]
I looked into it a little more, still could not figure our why, and also it behaves differently from the local machine and the jenkins.  At least for serial (probably also parallel), I could not reproduce the test failure in jenkins. 

The usage of the related to `cellZMinMax_` is pretty local, still not clear why it causes problem on Jenkins. 

With more comments and opinions, I think we can get this one in.  We might need to revisit it if more problem are revealed in the future.

**Comment by GitPaean** [comment]
jenkins build this update_data please

**Comment by bska** [comment]
> update_data please

You'll need to redo that once I merge OPM/opm-common#3021 and its associate update to the reference solutions, OPM/opm-tests#763.

**Comment by GitPaean** [comment]
> You'll need to redo that once I merge https://github.com/OPM/opm-common/pull/3021 and its associate update to the reference solutions, https://github.com/OPM/opm-tests/pull/763.

Sure. No problem.

**Comment by GitPaean** [comment]
jenkins build this update_data please

**Comment by GitPaean** [comment]
jenkins build this opm-tests=765 please

**Comment by GitPaean** [comment]
> We will have to revisit the problem before the next release though, because I'm not really comfortable having a system that produces different solutions when run in parallel versus when run in sequential mode.

As discussed above, this issue is there before this PR in the master branch already.  And I could not reproduces these failures locally.  Locally, there is no need to update the reference result at all.
